### PR TITLE
feat: New pattern `avm/ptn/authorization/resource-role-assignment`

### LINKED ISSUE #937: [AVM Module Issue]: Selective Role Assignment Modification for Existing Storage Accounts
### Check for previous/existing GitHub issues

- [x] I have checked for previous/existing GitHub issues

### Issue Type?

Feature Request

### Module Name

avm/res/storage/storage-account

### (Optional) Module Name if not listed above

_No response_

### (Optional) Module Version

_No response_

### Description

Description
Summary: Introduce a functionality within the storage account module (as well as others) that allows users to modify role assignments on existing storage accounts without requiring a full reconfiguration of the account settings. This feature is particularly essential for scenarios where the storage account is referenced by another service like Synapse, and role modifications are necessary post the initial creation.

Detailed Description:

Use Case: In cases where a storage account is created and subsequently referenced by a Synapse workspace, it becomes necessary to modify role assignments after the initial setup. Currently, altering role assignments requires calling the entire storage account module, involving reconfiguration of all settings, which is cumbersome and inefficient.
Proposed Solution: Implement a feature that enables calling the storage module with an option to solely update role assignments. This would streamline the process by allowing users to add or change roles without having to input all the other configurations again.
Expected Behavior:
Users should be able to call the storage module with a new parameter or method focused exclusively on role management.
This call would enable the addition, removal, or modification of roles associated with the storage account.
All other existing configurations of the storage account should remain unchanged unless explicitly specified by the user.
Benefits:
Enhanced flexibility and efficiency in managing role assignments for storage accounts.
Reduced risk of inadvertent configuration changes during role updates.
Better integration with services like Synapse, accommodating dynamic changes in role assignments post-initial setup.
Impact:

This feature would be a significant improvement for developers and administrators who manage cloud resources and require dynamic role management for storage accounts, especially in complex setups involving services like Synapse.

### (Optional) Correlation Id

_No response_

### PR DESCRIPTION
**Pull request by peterbud**
## Description

Adding new pattern `avm/ptn/authorization/resource-role-assignment`

https://github.com/Azure/Azure-Verified-Modules/issues/939

Migration of `modules/authorization/resource-scope-role-assignment`

Closes #937

## Pipeline Reference

| Pipeline |
| -------- |
| [avm.ptn.authorization.resource-role-assignment](https://github.com/peterbud/bicep-registry-modules/actions/workflows/avm.ptn.authorization.resource-role-assignment.yml) |

## Type of Change

- [ ] Update to CI Environment or utlities (Non-module effecting changes)
- [ ] Azure Verified Module updates:
  - [ ] Bugfix containing backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

### CODE REVIEW

File: avm/ptn/authorization/resource-role-assignment/main.bicep
Comment (AlexanderSehr) [inline avm/ptn/authorization/resource-role-assignment/main.bicep:81]:
```suggestion
        value: contains(roleDefinitionId, '/providers/Microsoft.Authorization/roleDefinitions/') ? roleDefinitionId : subscriptionResourceId('Microsoft.Authorization/roleDefinitions', roleDefinitionId)
```

Just a small helper that we allow in other modules deploying role assignments. Allows you to just specify the GUID as opposed to the full definition Id if preferred.

File: avm/ptn/authorization/resource-role-assignment/main.bicep
Comment (AlexanderSehr) [inline avm/ptn/authorization/resource-role-assignment/main.bicep:60]: Hey @peterbud would you happen to a have a link to a failed workflow run available? I'd like to investigate this further. Naturally, we don't need to block the PR over it, but it sounds like something we should fix in the tests...

File: avm/ptn/authorization/resource-role-assignment/main.bicep
Comment (AlexanderSehr) [inline avm/ptn/authorization/resource-role-assignment/main.bicep:66]: Just to align with the role assignments of other modules
```suggestion
  name: guid(resourceId, principalId, contains(roleDefinitionId, '/providers/Microsoft.Authorization/roleDefinitions/') ? roleDefinitionId : subscriptionResourceId('Microsoft.Authorization/roleDefinitions', roleDefinitionId))
```

Comment (AlexanderSehr) [inline avm/ptn/authorization/resource-role-assignment/main.bicep:66]:
```suggestion
  name: '${guid(resourceId, principalId, roleDefinitionId)}-ResourceRoleAssignment'
```
Just in case ;)

File: avm/ptn/authorization/resource-role-assignment/main.bicep
Comment (AlexanderSehr) [inline avm/ptn/authorization/resource-role-assignment/main.bicep:78]:
```suggestion
        value: !empty(name) ? name : guid(resourceId, principalId, contains(roleDefinitionId, '/providers/Microsoft.Authorization/roleDefinitions/') ? roleDefinitionId : subscriptionResourceId('Microsoft.Authorization/roleDefinitions', roleDefinitionId))
```

In alignment with how other modules do role assignments

File: avm/ptn/authorization/resource-role-assignment/main.bicep
Comment (AlexanderSehr) [inline avm/ptn/authorization/resource-role-assignment/main.bicep:9]:
```suggestion
@sys.description('Optional. The unique guid name for the role assignment.')
param name string?
```

Comment (AlexanderSehr) [reply avm/ptn/authorization/resource-role-assignment/main.bicep:9]: Note, for this and all other changes, I'd recommend to re-run `Set-AvmModule -ModuleFolderPath '.\avm\ptn\authorization\resource-role-assignment' -SkipFileAndFolderSetup -ThrottleLimit 5 -Recurse`

File: avm/ptn/authorization/resource-role-assignment/tests/e2e/defaults/main.test.bicep
Comment (AlexanderSehr) [inline avm/ptn/authorization/resource-role-assignment/tests/e2e/defaults/main.test.bicep:56]:
```suggestion
```
Can be removed with suggested change in `main.bicep` for the `name` parameter

File: avm/ptn/authorization/resource-role-assignment/main.bicep
Comment (peterbud) [reply avm/ptn/authorization/resource-role-assignment/main.bicep:60]: Hi @AlexanderSehr , here is the [failed workflow](https://github.com/peterbud/bicep-registry-modules/actions/runs/9024003041/job/24797001028#step:4:190)

Also I have noticed one similar case, when `loadJsonContent' creates a simelar non camel case variable here at the [storage-account/file-service](https://github.com/Azure/bicep-registry-modules/blob/dcd987fd8211f5dcd762eccd3c9ef3791e104838/avm/res/storage/storage-account/file-service/share/main.json#L227)

Comment (AlexanderSehr) [reply avm/ptn/authorization/resource-role-assignment/main.bicep:60]: Much appreciated, thanks @peterbud. I'll add it to my todo list. 
And ah right, thanks for reminding me of this horrible `fileShares` workaround I added some time ago. Almost successfully suppressed it from my memory 😄 Interesting though, that the static tests seemed not to have failed for the same/similar variable there though.

Anyways, will resolve this comment and make a note to investigate.

File: avm/ptn/authorization/resource-role-assignment/main.bicep
Comment (AlexanderSehr) [reply avm/ptn/authorization/resource-role-assignment/main.bicep:9]: > Note, for this and all other changes, I'd recommend to re-run `Set-AvmModule -ModuleFolderPath '.\avm\ptn\authorization\resource-role-assignment' -SkipFileAndFolderSetup -ThrottleLimit 5 -Recurse`

just bubbling this up in its own comment for awareness

**Comment by AlexanderSehr** [comment]
> Note, for this and all other changes, I'd recommend to re-run `Set-AvmModule -ModuleFolderPath '.\avm\ptn\authorization\resource-role-assignment' -SkipFileAndFolderSetup -ThrottleLimit 5 -Recurse`

just bubbling this up in its own comment for awareness

File: avm/ptn/authorization/resource-role-assignment/main.bicep
Comment (peterbud) [reply avm/ptn/authorization/resource-role-assignment/main.bicep:9]: I have a very strange consequence of making the `name` optional. Once I make it optional, the bicep linter gives an error later down the file:
`
The symbol "tFile" is declared in the context of the outer deployment and cannot be accessed by expressions within a nested deployment template that uses inner scoping for expression evaluation.
`

If I remove the optional `?` decorator from name, it works again.

Any suggestion?

Comment (peterbud) [reply avm/ptn/authorization/resource-role-assignment/main.bicep:9]: Should I make it required with a default value of empty string `''`?

Comment (AlexanderSehr) [reply avm/ptn/authorization/resource-role-assignment/main.bicep]: Hey @peterbud, sorry for the late reply (I'm actually out of office), but I'll see to clone your repository to try and replicate the behavior. The 2 subjects seem not to overlap. I'll let you know once I get around to it.

Comment (AlexanderSehr) [reply avm/ptn/authorization/resource-role-assignment/main.bicep]: Hey @peterbud,
I was able to reproduce the issue (which does not make any sense to me and may be a language bug). In any case, I did use your fork to implement an update to the Pester tests that you could find here: https://github.com/Azure/bicep-registry-modules/pull/1917

If it's merged sooner than later, I'd suggest to go with the original plan, which is
- Make the `name` variable nullable
- Remove the introduced explicit variable and invoke the function where the value is used directly

I'll ask the team to please fasttrack the review of the linked PR.

Comment (peterbud) [reply avm/ptn/authorization/resource-role-assignment/main.bicep]: Thanks, good news! Once your PR is merged, I'll move ahead with the suggested solution.

Comment (peterbud) [reply avm/ptn/authorization/resource-role-assignment/main.bicep]: Hi @AlexanderSehr , the updated Pester tests are working fine, and there is no test case error now with `LoadJsonContent`. 

However making the name optional and adding a handling of null `name` in the template assignment with formulas failed at deployment with a strange error message:

`Status Message: Deployment template validation failed: 'Template language version '2.0' does not support 'Outer' expression evaluation. Ensure you have set expressionEvaluationOptions.scope to 'Inner'.'. (Code:InvalidTemplate)`

More [here](https://github.com/peterbud/bicep-registry-modules/actions/runs/9066384300/job/24909256879#step:4:551)

I have tried different approaches, at the end which was working was to assign a default value to the `name` param - see the last commit. This way this param is still optional. Let me know if that's OK, or you have a better alternative.

File: avm/ptn/authorization/resource-role-assignment/main.bicep
Comment (AlexanderSehr) [inline avm/ptn/authorization/resource-role-assignment/main.bicep:1]:
```suggestion
metadata name = 'Resource-scoped role assignment'
```

Just thinking this renaming would make sense as 'all scopes' is technically not possible (as you cannot deploy this module to e.g. a management group scope). Just a minor detail.

After updating please run the `Set-AVMModule` script again.

Comment (AlexanderSehr) [reply avm/ptn/authorization/resource-role-assignment/main.bicep:1]: That's my last comment. Everything else looks good from my end. The most one could argue is that you presumably don't need the `!` at the `name` output anymore. But that's a nitpick so ... yeah 😄

**Comment by matebarabas** [comment]
@peterbud, @AlexanderSehr, please note that this module has just been filed for addition into the MAR/MCR manifest. Please do not attempt to publish this module, until its addition is confirmed (I'll let you know when the MCR PR is merged). Thank you!

File: avm/ptn/authorization/resource-role-assignment/main.bicep
Comment (AlexanderSehr) [inline avm/ptn/authorization/resource-role-assignment/main.bicep:78]: Hey @peterbud, how come you implemented this change? Should the parameter's default not be enough?

Comment (peterbud) [reply avm/ptn/authorization/resource-role-assignment/main.bicep:78]: I thought you meant this change when you have mentioned earlier:
> It's definitely ok with the !empty() check. 

Maybe I misunderstood you - if you want I can revert it back to simple:

```bicep

value: name
```

Let me know your preference.

Comment (AlexanderSehr) [reply avm/ptn/authorization/resource-role-assignment/main.bicep:78]: Hey @peterbud, 
sorry I just may have done a poor job explaining myself 😄 But yes, I mean only to remove the `!` from the output - as it shouldn't have been required in your version of the code. But it didn't break anything either which is why I said it's a nitpick 😉

Comment (peterbud) [reply avm/ptn/authorization/resource-role-assignment/main.bicep:78]: No worries, then I revert back to `value: name`

**Comment by matebarabas** [comment]
> @peterbud, @AlexanderSehr, please note that this module has just been filed for addition into the MAR/MCR manifest. Please do not attempt to publish this module, until its addition is confirmed (I'll let you know when the MCR PR is merged). Thank you!

@peterbud, @AlexanderSehr, this module has now been added to the MAR manifest. Please feel free to proceed with the review/publication as per the standard process. Thanks!

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**
